From 49410a9432e82e226b89cc564effb0c9a0bd0563 Mon Sep 17 00:00:00 2001
From: Jason Jin <Jason.jin@freescale.com>
Date: Mon, 19 Apr 2010 15:29:51 +0800
Subject: [PATCH 12/16] Make the FEC get mac addr from the u-boot setting.

Try to make the FEC get mac from the FEC register. The FEC
register should be initilized in u-boot.

Signed-off-by: Jason Jin <Jason.jin@freescale.com>
---
 arch/m68k/coldfire/m5441x/fec.c |   12 ++++--------
 drivers/net/fec.c               |   11 ++++++-----
 2 files changed, 10 insertions(+), 13 deletions(-)

diff --git a/arch/m68k/coldfire/m5441x/fec.c b/arch/m68k/coldfire/m5441x/fec.c
index 416aa43..2fc9801 100644
--- a/arch/m68k/coldfire/m5441x/fec.c
+++ b/arch/m68k/coldfire/m5441x/fec.c
@@ -112,19 +112,15 @@ static void fec_get_mac(struct net_device *dev)
 		* If it is all 1's or 0's, use the default.
 		*/
 		iap = FEC_FLASHMAC;
-		if ((iap[0] == 0) && (iap[1] == 0) && (iap[2] == 0) &&
-			(iap[3] == 0) && (iap[4] == 0) && (iap[5] == 0))
-			iap = fec_mac_default;
-		if ((iap[0] == 0xff) && (iap[1] == 0xff) &&
-			(iap[2] == 0xff) && (iap[3] == 0xff) &&
-			(iap[4] == 0xff) && (iap[5] == 0xff))
-			iap = fec_mac_default;
 	} else {
 		*((unsigned long *) &tmpaddr[0]) = fecp->fec_addr_low;
 		*((unsigned short *) &tmpaddr[4]) = (fecp->fec_addr_high >> 16);
-		iap = fec_mac_default;
+		iap = &tmpaddr[0];
 	}
 
+	if(!is_valid_ether_addr(iap))
+		iap = fec_mac_default;
+
 	memcpy(dev->dev_addr, iap, ETH_ALEN);
 
 	/* Adjust MAC if using default MAC address */
diff --git a/drivers/net/fec.c b/drivers/net/fec.c
index 28280ad..2d3f281 100644
--- a/drivers/net/fec.c
+++ b/drivers/net/fec.c
@@ -879,11 +879,6 @@ int __init fec_enet_init(struct net_device *dev,
 	fep->phy_hwp = fecp;
 #endif
 
-	/* Whack a reset.  We should wait for this.
-	*/
-	fecp->fec_ecntrl = 1;
-	udelay(10);
-
 	/* Set the Ethernet address.  If using multiple Enets on the 8xx,
 	 * this needs some work to get unique addresses.
 	 *
@@ -893,6 +888,12 @@ int __init fec_enet_init(struct net_device *dev,
 	if (plat && plat->get_mac)
 		plat->get_mac(dev);
 
+	/* Whack a reset.  We should wait for this.
+	*/
+	fecp->fec_ecntrl = 1;
+	udelay(10);
+
+
 	cbd_base = (cbd_t *)mem_addr;
 	/* XXX: missing check for allocation failure */
 	if (plat && plat->uncache)
-- 
1.6.4

